File: src/gobby/code_index/hasher.py
"""Content hashing for incremental indexing."""

from __future__ import annotations

import hashlib
from pathlib import Path


def file_content_hash(path: Path | str) -> str:
    """SHA-256 hash of entire file contents."""
    h = hashlib.sha256()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            h.update(chunk)
    return h.hexdigest()


def symbol_content_hash(source: bytes, start: int, end: int) -> str:
    """SHA-256 hash of a byte slice (symbol source)."""
    return hashlib.sha256(source[start:end]).hexdigest()

File: rust/gcode/src/index/hasher.rs
//! Content hashing for incremental indexing.
//! Ports logic from src/gobby/code_index/hasher.py.

use sha2::{Digest, Sha256};
use std::io::Read;
use std::path::Path;

/// SHA-256 hash of entire file contents.
pub fn file_content_hash(path: &Path) -> anyhow::Result<String> {
    let mut file = std::fs::File::open(path)?;
    let mut hasher = Sha256::new();
    let mut buf = [0u8; 65536];
    loop {
        let n = file.read(&mut buf)?;
        if n == 0 {
            break;
        }
        hasher.update(&buf[..n]);
    }
    Ok(format!("{:x}", hasher.finalize()))
}

/// SHA-256 hash of a byte slice (symbol source).
pub fn symbol_content_hash(source: &[u8], start: usize, end: usize) -> String {
    let mut hasher = Sha256::new();
    hasher.update(&source[start..end]);
    format!("{:x}", hasher.finalize())
}
